### DISCUSSION TITLE
Alias types (Strongly typed IDs?) formalized

### DISCUSSION THREAD

**Discussion by roa-nyx** [dotnet/csharplang#7094] [Language Ideas]
First time contributor here, sorry if I'm doing anything out of order.

I had a thought that might be a pretty cool feature for C# (assuming it doesn't exist already?). I feel like this could probably be done with source code generators, but my limited experience in designing one has scared me off of them.

I'm looking for discussion, suggestions on how this could be done currently, and improvements, or just reasons why this could never work.

# My Proposal

I'd like to propose a form of "alias" types. I've seen others refer to similar schemes as "Strongly Typed IDs" but I feel that's too restrictive a name, as there is no need for it to be relegated only to IDs at all in my mind, but they are a simple starting point for discussion.

So my thought is something along these lines:

```csharp
//aliastype underlyingType:aliasName;
aliastype string:SomeID;
```

And that's it! Everywhere where `SomeID` is used as a type it would effectively just lower it to it's underlying type:

```csharp
public void INeedSomeID(SomeID id){
}
```
would lower to:
```csharp
public void INeedSomeID(string id){
}
```

Where I think this is useful is in it's static analysis capabilities if we add the following rules:

- An alias type can be implicitly/explicitly cast to it's underlying type
- The underlying type can not be implicitly cast to it's alias type
- Two alias types, regardless of underlying types, can not be implicitly or explicitly cast to one another

```csharp
aliastype string:SomeID;
aliastype string:SomeOtherID;
public void INeedSomeID(SomeID id){
}
public void INeedAString(SomeID id){
}

string idRaw = "id123";
SomeID id = new SomeID(idRaw);
//C# could have a "default" constructor implementation for alias types, which would simply be lowered to the single parameter:
//string id = idRaw;
SomeOtherID otherId = (SomeOtherID)"id456";

INeedSomeID(id); //All good!
INeedSomeID(idRaw); //Error, you must explicitly cast the string to SomeID
INeedSomeID((SomeID)idRaw); //All good again!
INeedSomeID((SomeOtherID)idRaw); //Error, type mismatch
INeedSomeID((SomeID)(string)otherId); //All good!

INeedAString(idRaw); //All good!
INeedAString(id); //All good!
INeedAString((string)id); //All good!
INeedAString((SomeID)idRaw); //All good!
```

Ok so that's my proposal in a nutshell.

# A contrived example

Where I think this could be used, please consider:

```csharp
class Person{
    public readonly string Name; 
    //...
}
class Pet{
    public readonly string Name;
    //...
}

class PetRegistryService{
    public void RegisterPet(string ownerName, string petName){
        //..
    }
}
```

In the above scenario I can easily see myself making the mistake of calling the parameters in the wrong order. While being careful is of course always important I think that making it a little more explicit can be useful in some contexts.

With my proposal this could be written as:
```csharp
aliastype string:PersonName;
aliastype string:PetName;

class Person{
    public readonly PersonName Name;
}
class Pet{
    public readonly PetName Name;
}

class PetRegistryService{
    public void RegisterPet(PersonName ownerName, PetName petName){
        //..
    }
}
```

Now if anyone accidentally calls RegisterPet with two PetNames by accident the IDE/compiler could emit an error.

I think aliastypes would obey standard namespacing rules, and could be namespaced even if declared within classes (kinda like static variables are).

# Shorthands

I always like to take things further, so I came up with these two shorthands:

## Shorthand 1

Declare an alias type directly where the type is being used.

```csharp
class Person{
    public readonly string:Name Name;
}
class Pet{
    public readonly string:Name Name;
}

class PetRegistryService{
    public void RegisterPet(Person.Name ownerName, Pet.Name petName){
        //..
    }
}
```

Sweet now we don't need the `aliastype` directive except where it doesn't make sense for the declaration to exist in a single type/class.

## Shorthand 2

Declare an alias type directly where the type is being used, and implicitly use the alias name as the member name.

```csharp
class Person{
    public readonly string:Name;
}
class Pet{
    public readonly string:Name;
}

class PetRegistryService{
    public void RegisterPet(Person.Name ownerName, Pet.Name petName){
        //..
    }
}
```

In the above both Person and Pet would still be equivalent to shorthand 1. We just don't need to repeat "Name" twice if we want a member called the alias type's name.

# Shortcomings of current workarounds

How I can imagine this being "solved" currently is something like this (there is probably something better):

```csharp
class Person{
    public readonly PersonName Name;
    //I'm using record here for their equality checking and immutability behaviour
    //I'm using a struct to avoid an extra level of unnecessary redirect, and memory allocation
    record struct PersonName(string Name){
        //I couldn't find any documentation on whether records can have implicit/explicit converters but I don't see why not, so I'm  making the assumption that's posssible.
        //implicit converter to string
        //explicit converter from string
    }
}
class Pet{
    public readonly PetName Name;
    record struct PetName(string Name){ /* ... */ }
}

class PetRegistryService{
    public void RegisterPet(Person.PersonName ownerName, Pet.PetName petName){
        //..
    }
}
```

In my mind the cons of this approach:
- SUPER verbose. We literally just want a basic type that has some additional type safety that can be implied without all this boilerplate.
- I used a struct above to avoid an extra level of redirection to get to the underlying value, but that has different nullability behaviour so may be undesired depending on the specifics of the situation.

I feel like my proposal solves all the cons without introducing any other cons that I can think of.

**Comment by HaloFour** [comment]
Check out "explicit extensions", I think they largely cover the use case of wanting a zero-cost typesafe alias for an underlying type without relying on inheritance:

https://github.com/dotnet/csharplang/blob/main/meetings/working-groups/roles/extensions-2023-02-21.md

So an alias might be:

```cd
public explicit extension PersonName for string { }
```

You could then explicitly cast a `string` to a `PersonName` and pass it around like a proper type, although under the hood it's still just a normal `string`.  Additional members could be added to `PersonName` as well.

**Comment by lsoft** [comment]
zero cost 'Strongly typed id's can be implemented via new `alias any type`.

Something lke this:

```C#
using strict PersonId = int; //any syntax is accepted here; I created a new keyword strict
using strict SessionId = int; //any syntax is accepted here; I created a new keyword strict

...

public class PersonStruct { PersonId Id; }
public class SessionStruct { SessionId Id; }

...

person.Id = session.Id; //error!
person.Id = (PersonId)(int)session.Id; //workaround if you 100% need it
```

there are disadvantages, of course. for example, it will work only in this file (`using global strict PersonId = int;`?)

anyway, zero cost strongly typed id is a really useful feature regardless of its implementation way. 👍
it would be cool to see such feature in future C#.

**Reply by CyrusNajmabadi** [reply]
This is already possible with `struct records`.  It's even simpler syntax: `struct record PersonId(int underlying)`.  You can also add conversion operators manually (or with an SG if you want) if you really want to be able to coerce like you showed.

**Reply by lsoft** [reply]
@CyrusNajmabadi yes, it is already possible, for example with Andew Lock's source generator...

I would like to know, what's the runtime cost of such approach? may be anyone tested this?

Also, is it 100% correct to get this feature at runtime? I guess not. this feature is a compile-time-guard feature.

**Reply by CyrusNajmabadi** [reply]
> I would like to know, what's the runtime cost of such approach? may be anyone tested this?

There is no runtime cost.  A struct containing a single item is the size of that item.

> Also, is it 100% correct to get this feature at runtime?

I'm not sure what this means.

**Comment by lsoft** [comment]
@CyrusNajmabadi my quick and dirty tests showed a difference in perf. I need to go deeper to IL or ASM, but here is too late...

If these results are correct and the differenct exists, it will make sense to raise a question: will C# benefit from compile-time `strongly typed id` feature? 'compile time' means there will be no traces in reflections and binaries. so `typeof(PersonStruct).GetField("Id").FieldType` will return `Int32`.

```
BenchmarkDotNet=v0.13.5, OS=Windows 10 (10.0.19045.2846/22H2/2022Update)
AMD Ryzen 7 4700U with Radeon Graphics, 1 CPU, 8 logical and 8 physical cores
.NET SDK=7.0.300-preview.23179.2
  [Host]     : .NET 6.0.16 (6.0.1623.17311), X64 RyuJIT AVX2
  Job-CHREBG : .NET 6.0.16 (6.0.1623.17311), X64 RyuJIT AVX2

Runtime=.NET 6.0  IterationCount=100

|               Method |     Mean |     Error |    StdDev | Allocated |
|--------------------- |---------:|----------:|----------:|----------:|
|     Use_IntContainer | 2.640 ns | 0.0157 ns | 0.0462 ns |         - |
| Use_StructContainer1 | 3.056 ns | 0.0151 ns | 0.0446 ns |         - |
| Use_StructContainer2 | 3.046 ns | 0.0176 ns | 0.0508 ns |         - |
```

```C#
    [SimpleJob(RuntimeMoniker.Net60, iterationCount: 100)]
    [MemoryDiagnoser]
    public class Fixture
    {
        [Benchmark]
        public bool Use_IntContainer()
        {
            var result = false;

            for (var i = 0; i < 10; i++)
            {
                var c1 = new IntContainer(1);
                result |= 50 == c1.Id;
            }

            return result;
        }

        [Benchmark]
        public bool Use_StructContainer1()
        {
            var result = false;

            for (var i = 0; i < 10; i++)
            {
                var c1 = new StructContainer(new StructId(1));
                result |= 50 == c1.Id;
            }

            return result;
        }

        [Benchmark]
        public bool Use_StructContainer2()
        {
            var result = false;

            for (var i = 0; i < 10; i++)
            {
                var c1 = new StructContainer(new StructId(1));
                result |= c1.Id == 50;
            }

            return result;
        }
    }


    public readonly struct IntContainer
    {
        public readonly int Id;

        public IntContainer(int id)
        {
            Id = id;
        }
    }

    public readonly struct StructContainer
    {
        public readonly StructId Id;

        public StructContainer(StructId id)
        {
            Id = id;
        }
    }

    public record struct StructId(int Id)
    {
        public static implicit operator int(StructId d) => d.Id;
    }

```

**Reply by HaloFour** [reply]
I wouldn't trust any benchmark that only shows a difference of 400 picoseconds.  I think the benchmark would have to do a lot more to reliably demonstrate a difference.  This is also using multiple wrappers.  I can't speak to how deep the runtime can optimize those scenarios, but if you were to compare the native code for `int` vs. a simple struct wrapping a single `int` the resultant native code is identical.

Given:

```cs
public record struct Wrapped(int value);

public static class C {
    public static int M1(int x, int y) {
        return x + y;
    }
    
    public static int M2(Wrapped x, Wrapped y) {
        return x.value + y.value;
    }
}
```

Result:

```
C.M1(Int32, Int32)
    L0000: lea eax, [ecx+edx]
    L0003: ret

C.M2(Wrapped, Wrapped)
    L0000: lea eax, [ecx+edx]
    L0003: ret
```

**Reply by CyrusNajmabadi** [reply]
> my quick and dirty tests showed a difference in perf

File issue with runtime.  These should be erased (as Halo's example shows as well).
